examples/contrib/oed/guides/amort.py

```diff
@@ -0,0 +1,152 @@
+        xtxi = rinverse(xtx, sym=True)
+        mu = rmv(xtxi, rmv(design.transpose(-1, -2), y))
+
+        # Extract sub-indices
+        mu = tensor_to_dict(self.w_sizes, mu, subset=target_labels)
+        scale_tril = {l: rtril(self.scale_tril[l]) for l in target_labels}
+
+        return mu, scale_tril
+
+    def guide(self, y_dict, design, observation_labels, target_labels):
+
+        pyro.module("ba_guide", self)
+
+        # Returns two dicts from labels -> tensors
+        mu, scale_tril = self.forward(y_dict, design, target_labels)
+
+        for l in target_labels:
+            w_dist = dist.MultivariateNormal(mu[l], scale_tril=scale_tril[l])
+            pyro.sample(l, w_dist)
+
+
+class SigmoidGuide(LinearModelGuide):
+
+    def __init__(self, d, n, w_sizes):
+        super(SigmoidGuide, self).__init__(d, w_sizes)
+        self.inverse_sigmoid_scale = nn.Parameter(torch.ones(n))
+        self.h1_weight = nn.Parameter(torch.ones(n))
+        self.h1_bias = nn.Parameter(torch.zeros(n))
+
+    def forward(self, y_dict, design, target_labels):
+
+        y = torch.cat(list(y_dict.values()), dim=-1)
+
+        # Approx invert transformation on y in expectation
+        y, y1m = y.clamp(1e-35, 1), (1.-y).clamp(1e-35, 1)
+        logited = y.log() - y1m.log()
+        y_trans = logited/.1
+        y_trans = y_trans * self.inverse_sigmoid_scale
+        hidden = self.softplus(y_trans)
+        y_trans = y_trans + hidden * self.h1_weight + self.h1_bias
+
+        return self.linear_model_formula(y_trans, design, target_labels)
+
+
+class NormalInverseGammaGuide(LinearModelGuide):
+
+    def __init__(self, d, w_sizes, mf=False, tau_label="tau"):
+        super(NormalInverseGammaGuide, self).__init__(d, w_sizes)
+        self.alpha = nn.Parameter(100.*torch.ones(d))
+        self.b0 = nn.Parameter(100.*torch.ones(d))
+        self.mf = mf
+        self.tau_label = tau_label
+
+    def forward(self, y_dict, design, target_labels):
+
+        y = torch.cat(list(y_dict.values()), dim=-1)
+
+        coefficient_labels = [label for label in target_labels if label != self.tau_label]
+        mu, scale_tril = self.linear_model_formula(y, design, coefficient_labels)
+        mu_vec = torch.cat(list(mu.values()), dim=-1)
+
+        yty = rvv(y, y)
+        ytxmu = rvv(y, rmv(design, mu_vec))
+        beta = self.b0 + .5*(yty - ytxmu)
+
+        return mu, scale_tril, self.alpha, beta
+
+    def guide(self, y_dict, design, observation_labels, target_labels):
+
+        pyro.module("ba_guide", self)
+
+        mu, scale_tril, alpha, beta = self.forward(y_dict, design, target_labels)
+
+        if self.tau_label in target_labels:
+            tau_dist = dist.Gamma(alpha, beta)
+            tau = pyro.sample(self.tau_label, tau_dist)
+            obs_sd = 1./tau.sqrt().unsqueeze(-1).unsqueeze(-1)
+
+        for label in target_labels:
+            if label != self.tau_label:
+                if self.mf:
```

Comment:
If mean field, posterior is Gaussian and independent of `tau`. Otherwise they are coupled in the correct NIG way

Reply:
Not clear what would happen if tau was not a variable of interest but you wanted non-mean-field